File: smallestPositiveInteger.java
/*

Write a function:

class Solution { public int solution(int[] A); }

that, given an array A of N integers, returns the smallest positive integer (greater than 0) that does not occur in A.

For example, given A = [1, 3, 6, 4, 1, 2], the function should return 5.

Given A = [1, 2, 3], the function should return 4.

Given A = [−1, −3], the function should return 1.

Assume that:

N is an integer within the range [1..100,000];
each element of array A is an integer within the range [−1,000,000..1,000,000].
Complexity:

expected worst-case time complexity is O(N);
expected worst-case space complexity is O(N) (not counting the storage required for input arguments).

*/

// you can also use imports, for example:
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int[] A) {
        // write your code in Java SE 8
        
        Set<Integer> set = new HashSet<Integer>();
        for (int i = 0; i < A.length; i++) {
            if (A[i] >= 1 && !set.contains(A[i])) {
                set.add(A[i]);
            }
        }

        int smallestPositiveInteger = 1;        
        while (set.contains(smallestPositiveInteger)) {
            smallestPositiveInteger++;
        }
        
        return smallestPositiveInteger;
    }
}

File: smallestPositiveInteger.js
// you can write to stdout for debugging purposes, e.g.
// console.log('this is a debug message');

function solution(A) {
    // write your code in JavaScript (Node.js 8.9.4)
    
    let sorted = A.sort();
    let storage = {};
    
    // Store sorted Array in a json using the value as the key
    for (let i = 0; i < sorted.length; i++) {
        storage[sorted[i]] = sorted[i];
    }
    
    // If the storage does not contain the key from the loop below, return the key
    for (let i = 1; i < 1000001; i++) {
        if (!storage[i]) return i; 
    }
    
    return 1;
}